keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1899 @@
+  - pod eviction as a possible [future improvement](#kubelet-initiated-pod-eviction)
+- scheduler
+  - simple node filtering based on the availability of QoS-class resources in
+    the first implementation phase (with possible
+    [future improvements](#scheduler-improvements))
+  - pod priority based eviction is in effect
+- quota, limits and defaults are [future work](#future-work) items
+
+
+### CRI API
+
+The following additions to the CRI protocol are suggested.
+
+#### ContainerConfig
+
+The `ContainerConfig` message will be supplemented with new `class_resources`
+field, providing per-container setting for QoS-class resources. This will be
+used in `CreateContainerRequest` to communicate the container-level QoS-class
+resource assignments to the runtime.
+
+```diff
+ message ContainerConfig {
+
+ ...
+     // Configuration specific to Linux containers.
+     LinuxContainerConfig linux = 15;
+     // Configuration specific to Windows containers.
+     WindowsContainerConfig windows = 16;
++
++    // Configuration of QoS resources.
++    ContainerQoSResources qos_resources = 17;
+ }
+
++// ContainerQoSResources specifies the configuration of QoS resources of a
++// container.
++message ContainerQoSResources {
++    // QoS resources the container will be assigned to.
++    // Key-value pairs where key is name of the QoS resource and value is the
++    // name of the class.
++    map<string, string> classes = 1;
++}
+```
+
+#### UpdateContainerResourcesRequest
+
+Similar to `CreateContainerRequest`, the `UpdateContainerResourcesRequest`
+message will extended to allow updating of QoS-class resource configuration of
+a running container.  Depending on runtime-level support of a particular
+resource (and possibly the type of resource) UpdateContainerResourcesRequest
+might fail. Resource discovery (see [Runtime status](#runtime-status)) has
+the capability to distinguish immutable resource types.
+
+Note that neither of the existing QoS-class resource types (RDT or blockio)
+support updates because of runtime limitations, yet.
+
+```diff
+ message UpdateContainerResourcesRequest {
+
+ ...
+     // resources to update or other options to use when updating the container.
+     map<string, string> annotations = 4;
++    // Configuration of QoS resources.
++    ContainerQoSResources qos_resources = 5;
+}
+```
+
+#### PodSandboxConfig
+
+The `PodSandboxConfig` will be supplemented with a new `class_resources` field
+that specifies the assignment of pod-level QoS-class resources. The intended
+use for this would be to be able to communicate pod-level QoS-class resource
+assignments at sandbox creation time (`RunPodSandboxRequest`).
+
+```diff
+ message PodSandboxConfig {
+@@ -45,5 +45,14 @@ message PodSandboxConfig {
+     LinuxPodSandboxConfig linux = 8;
+     // Optional configurations specific to Windows hosts.
+     WindowsPodSandboxConfig windows = 9;
++    // Configuration of QoS resources.
++    PodQoSResources qos_resources = 10;
```

Comment:
Could make this repeated and have a strong type with resource name and class value instead.

Reply:
Like this?

```diff
...
+    // Configuration of QoS resources.
+    repeated PodQoSResource qos_resources = 10;
 }

+message PodQOSResource {
+    string name = 1;                                           
+    string class = 2;                                           
+} 


```


and dittor for container qos resources

Reply:
@mrunalp this has been changed, for Pod sandbox:

```diff
 message PodSandboxConfig {
...
     WindowsPodSandboxConfig windows = 9;
+    repeated PodQOSResource qos_resources = 10;
 }

+message PodQOSResource {
+    string name = 1;
+    string class = 2;
+}
```

and similarly for the container

```diff
 message ContainerConfig {
 ...
     WindowsContainerConfig windows = 16;
+    repeated ContainerQOSResource qos_resources = 17;
 }

+message ContainerQOSResource {
+    string name = 1;
+    string class = 2;
+}

Reply:
okay :+1: